internals/overlord/logstate/loki/loki.go

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package loki
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/cmd"
+	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+const maxRequestEntries = 1000
+
+var requestTimeout = 10 * time.Second
+
+type Client struct {
+	targetName string
+	remoteURL  string
+	// buffered entries are "sharded" by service name
+	entries map[string][]lokiEntry
+	// Keep track of the number of buffered entries, to avoid having to iterate
+	// the entries map to get the total number.
+	numEntries int
+
+	httpClient *http.Client
+}
+
+func NewClient(target *plan.LogTarget) *Client {
+	return &Client{
+		targetName: target.Name,
+		remoteURL:  target.Location,
+		entries:    map[string][]lokiEntry{},
+		httpClient: &http.Client{Timeout: requestTimeout},
+	}
+}
+
+func (c *Client) Write(ctx context.Context, entry servicelog.Entry) error {
+	c.entries[entry.Service] = append(c.entries[entry.Service], encodeEntry(entry))
+	c.numEntries++
+
+	if c.numEntries >= maxRequestEntries {
+		return c.Flush(ctx)
```

Comment:
Per previous review, it seems our API should have `Write` meaning "buffer data" until `Flush`, which is controlled by the central system. Why would we ask every client to control the number of entries before flush, when we can do that just once from the place that does both `Write` and `Flush`? The entries written are known there, and we already have a mechanism to `Flush` timely.

To be clear, maybe that's not a good idea for some other reason and won't survive the test of times, but it looks like from where we stand at the moment that's worth trying?

Reply:
The issue here is that the `logClient` interface may also be implemented by non-buffering clients. E.g. in the case of syslog, the client will just keep an open TCP/UDP connection and send log entries as they are received - no buffering.

Unless I'm mistaken, what you're suggesting is to keep track of the number of buffered logs **in the gatherer,** then call `Flush` on the client when we cross the threshold. This doesn't really make sense for non-buffering clients though. So it would require the gatherer to know some internal details about the client (does it buffer or not?), which defeats the purpose of having a `logClient` interface.

`Write` should really be understood as "write and possibly flush", while `Flush` should be understood as "force flush". This is well-documented on the `logClient` interface - however I'm happy to change these method names if you don't think they accurately reflect the semantics here.

I agree it feels kind of wrong to be calling `Flush` inside the `Write` method, but at the end of the day, this is an implementation detail. The gatherer should treat the client as a black box where it sends logs, and shouldn't take an opinion on implementation such as this.